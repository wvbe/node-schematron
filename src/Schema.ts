import { evaluateXPath } from 'fontoxpath';
import { sync } from 'slimdom-sax-parser';

import { Variable, JsonVariable } from './Variable';
import { Phase } from './Phase';
import { Pattern } from './Pattern';
import { Namespace } from './Namespace';
import { Result } from './Result';

export class Schema {
	public title: string;
	public defaultPhase: string | null;
	public variables: Variable[];
	public phases: Phase[];
	public patterns: Pattern[];
	public namespaces: Namespace[];

	constructor(
		title: string,
		defaultPhase: string | null,
		variables: Variable[],
		phases: Phase[],
		patterns: Pattern[],
		namespaces: Namespace[]
	) {
		this.title = title;
		this.defaultPhase = defaultPhase;
		this.variables = variables;
		this.phases = phases;
		this.patterns = patterns;
		this.namespaces = namespaces;
	}

	validateString(documentXmlString: string, phaseId?: string): Result[] {
		// Typescript casting slimdom.Document to Document, which are the same
		return this.validateDocument((sync(documentXmlString) as unknown) as Document, phaseId);
	}

	validateDocument(documentDom: Document, phaseId?: string): Result[] {
		if (!phaseId) {
			phaseId = '#DEFAULT';
		}
		if (phaseId === '#DEFAULT') {
			phaseId = this.defaultPhase || '#ALL';
		}

		const namespaceResolver = this.getNamespaceUriForPrefix.bind(this);
		const variables = Variable.reduceVariables(
			documentDom,
			this.variables,
			namespaceResolver,
			{}
		);

		if (phaseId === '#ALL') {
			return this.patterns.reduce(
				(results: Result[], pattern) =>
					results.concat(
						pattern.validateDocument(documentDom, variables, namespaceResolver)
					),
				[]
			);
		}

		const phase = this.phases.find(phase => phase.id === phaseId);
		const phaseVariables = Variable.reduceVariables(
			documentDom,
			phase?.variables || [],
			namespaceResolver,
			{
				...variables
			}
		);

		return (
			phase?.active
				.map(patternId => this.patterns.find(pattern => pattern.id === patternId))
				.reduce(
					(results: Result[], pattern) =>
						results.concat(
							pattern?.validateDocument(
								documentDom,
								phaseVariables,
								namespaceResolver
							) || []
						),
					[]
				) || []
		);
	}

	// TODO more optimally store the namespace prefix/uri mapping. Right now its modeled as an array because there
	// is a list of <ns> elements that are not really guaranteed to use unique prefixes.
	getNamespaceUriForPrefix(prefix?: string | null): string | null {
		if (!prefix) {
			return null;
		}
		const ns = this.namespaces.find(ns => ns.prefix === prefix);
		if (!ns) {
			throw new Error(
				`Namespace prefix "${prefix}" could not be resolved to an URI using <sch:ns>`
			);
		}

		return ns.uri;
	}

	static QUERY = `
		declare namespace sch = 'http://purl.oclc.org/dsdl/schematron';

		declare function local:json($node as node()) {
			if ($node[self::text()])
				then $node/string()
			else
			map:merge((
				map:entry('$type', $node/local-name()),
				for $attr in $node/@*
					return map:entry($attr/name(), $attr/string())
			))
		};

		let $context := /*[1]
		return map {
			'title': $context/@title/string(),
			'defaultPhase': $context/@defaultPhase/string(),
			'phases': array { $context/sch:phase/${Phase.QUERY}},
			'patterns': array { $context/sch:pattern/${Pattern.QUERY}},
			'variables': array { $context/sch:let/${Variable.QUERY}},
			'namespaces': array { $context/sch:ns/${Namespace.QUERY}}
		}
	`;

	static fromJson(json: JsonSchema): Schema {
		return new Schema(
			json.title,
			json.defaultPhase,
			json.variables.map(obj => Variable.fromJson(obj)),
			json.phases.map(obj => Phase.fromJson(obj)),
			json.patterns.map(obj => Pattern.fromJson(obj)),
			json.namespaces.map(obj => Namespace.fromJson(obj))
		);
	}

	static fromDomToJson(schematronDom: Document): JsonSchema {
		return evaluateXPath(Schema.QUERY, schematronDom, null, {}, undefined, {
			language: evaluateXPath.XQUERY_3_1_LANGUAGE
		});
	}

	static fromDom(schematronDom: Document): Schema {
		return Schema.fromJson(Schema.fromDomToJson(schematronDom));
	}

	static fromString(schematronXmlString: string): Schema {
		return Schema.fromDom((sync(schematronXmlString) as unknown) as Document);
	}
}

export type JsonSchema = {
	title: string;
	defaultPhase: string | null;
	variables: JsonVariable[];
	phases: Phase[];
	patterns: Pattern[];
	namespaces: Namespace[];
};
