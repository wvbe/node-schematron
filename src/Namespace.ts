export class Namespace {
	prefix: string;
	uri: string;

	constructor(prefix: string, uri: string) {
		this.prefix = prefix;
		this.uri = uri;
	}

	static QUERY = `map {
		"prefix": @prefix/string(),
		"uri": @uri/string()
	}`;

	static fromJson(json: JsonNamespace): Namespace {
		return new Namespace(json.prefix, json.uri);
	}
}

export type JsonNamespace = {
	prefix: string;
	uri: string;
};
