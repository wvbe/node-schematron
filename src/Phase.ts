import { Variable } from './Variable';

export class Phase {
	id: string;
	active: string[];
	variables: Variable[];

	constructor(id: string, active: string[], variables: Variable[]) {
		this.id = id;
		this.active = active;
		this.variables = variables;
	}

	static QUERY = `map {
		"id": @id/string(),
		"active": array { ./sch:active/@pattern/string() },
		'variables': array { ./sch:let/${Variable.QUERY}}
	}`;

	static fromJson(json: JsonPhase): Phase {
		return new Phase(
			json.id,
			json.active,
			json.variables.map(rule => Variable.fromJson(rule))
		);
	}
}

export type JsonPhase = {
	id: string;
	active: string[];
	variables: Variable[];
};
