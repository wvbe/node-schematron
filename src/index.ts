// export const sum = (a: number, b: number) => {
//   if ('development' === process.env.NODE_ENV) {
//     console.log('boop');
//   }
//   return a + b;
// };

export { registerCustomXPathFunction } from 'fontoxpath';

export { Schema } from './Schema';
