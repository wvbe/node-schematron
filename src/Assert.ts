import { evaluateXPathToBoolean, evaluateXPathToString } from 'fontoxpath';

import { Result } from './Result';

import { FontoxpathOptions } from './types';

export class Assert {
	id: string | null;
	test: string;
	message: Array<string | Object>;
	isReport: boolean;

	constructor(
		id: string | null,
		test: string,
		message: Array<string | Object>,
		isReport: boolean
	) {
		this.id = id;
		this.test = test;
		this.message = message;
		this.isReport = isReport;
	}

	createMessageString(
		contextNode: Node,
		variables: Object,
		fontoxpathOptions: FontoxpathOptions,
		chunks: Array<string | any>
	): string {
		return chunks
			.map((chunk): string => {
				if (typeof chunk === 'string') {
					return chunk;
				}

				// <sch:name />
				if (chunk.$type === 'name') {
					return evaluateXPathToString(
						'name(' + (chunk.path || '') + ')',
						contextNode,
						null,
						variables,
						fontoxpathOptions
					);
				}

				// <sch:value-of />
				if (chunk.$type === 'value-of') {
					return evaluateXPathToString(
						chunk.select,
						contextNode,
						null,
						variables,
						fontoxpathOptions
					);
				}

				console.log(chunk);
				throw new Error('Unsupported element in <sch:message>');
			})
			.join('');
	}

	validateNode(
		context: Node,
		variables: Object,
		fontoxpathOptions: FontoxpathOptions
	): Result | null {
		const outcome = evaluateXPathToBoolean(
			this.test,
			context,
			null,
			variables,
			fontoxpathOptions
		);
		return (!this.isReport && outcome) || (this.isReport && !outcome)
			? null
			: new Result(
					context,
					this,
					this.createMessageString(context, variables, fontoxpathOptions, this.message)
			  );
	}

	static QUERY = `map {
		'id': if (@id) then string(@id) else (),
		'test': @test/string(),
		'message': array { (./text()|./element())/local:json(.) },
		'isReport': boolean(local-name() = 'report')
	}`;

	static fromJson(json: AssertJson): Assert {
		return new Assert(json.id, json.test, json.message, json.isReport);
	}
}

export type AssertJson = {
	id: string | null;
	test: string;
	message: Array<string | Object>;
	isReport: boolean;
};
