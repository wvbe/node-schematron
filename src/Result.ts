import { Node, serializeToWellFormedString } from 'slimdom';
import { Assert } from './Assert';

export class Result {
	// pattern: Pattern;
	// phase?: Phase;
	// rule: Rule;
	assertId: string | null;
	isReport: boolean;
	context: Node;
	message?: string;

	constructor(
		// pattern: Pattern,
		// phase?: Phase,
		// rule: Rule,
		context: Node,
		assert: Assert,
		message?: string
	) {
		// this.pattern = pattern;
		// this.phase = phase;
		// this.rule = rule;
		this.assertId = assert.id;
		this.isReport = assert.isReport;
		this.context = context;
		this.message = message;
	}

	toJson() {
		return {
			assertId: this.assertId,
			isReport: this.isReport,
			context: serializeToWellFormedString(this.context),
			message: this.message
		};
	}
}
