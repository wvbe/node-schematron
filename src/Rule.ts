import { Variable, JsonVariable } from './Variable';
import { Assert, JsonAssert } from './Assert';
import { Result } from './Result';

export class Rule {
	context: string;
	variables: Variable[];
	asserts: Assert[];

	constructor(context: string, variables: Variable[], asserts: Assert[]) {
		this.context = context;
		this.variables = variables;
		this.asserts = asserts;
	}

	validateNode(
		context: Node,
		parentVariables: Object | null,
		namespaceResolver: (prefix?: string | null) => string | null
	): Result[] {
		const variables = Variable.reduceVariables(context, this.variables, namespaceResolver, {
			...parentVariables
		});

		return this.asserts
			.map(assert => assert.validateNode(context, variables, namespaceResolver))
			.filter(result => result !== null) as Result[];
	}

	static QUERY = `map {
		'context': @context/string(),
		'variables': array { ./sch:let/${Variable.QUERY}},
		'asserts': array{ ./(sch:report|sch:assert)/${Assert.QUERY}}
	}`;

	static fromJson(json: JsonRule): Rule {
		const variables = json.variables.map(rule => Variable.fromJson(rule));
		const asserts = json.asserts.map(rule => Assert.fromJson(rule));

		return new Rule(json.context, variables, asserts);
	}
}

export type JsonRule = {
	context: string;
	variables: JsonVariable[];
	asserts: JsonAssert[];
};
