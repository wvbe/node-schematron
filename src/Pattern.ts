import { evaluateXPathToNodes } from 'fontoxpath';

import { Variable, JsonVariable } from './Variable';
import { Rule, JsonRule } from './Rule';
import { Result } from './Result';

export class Pattern {
	id: string | null;
	rules: Rule[];
	variables: Variable[];

	constructor(id: string | null, rules: Rule[], variables: Variable[]) {
		this.id = id;
		this.rules = rules;
		this.variables = variables;
	}

	validateDocument(
		documentDom: Document,
		parentVariables: object | null,
		namespaceResolver: (prefix?: string | null) => string | null
	) {
		const variables = Variable.reduceVariables(documentDom, this.variables, namespaceResolver, {
			...parentVariables
		});
		const ruleContexts = this.rules.map(rule =>
			evaluateXPathToNodes('//(' + rule.context + ')', documentDom, null, variables, {
				namespaceResolver
			})
		);
		const flattenValidationResults = (results: Result[], node: Node): Result[] => {
			const ruleIndex = ruleContexts.findIndex(context => context.includes(node));
			const rule = ruleIndex >= 0 ? this.rules[ruleIndex] : null;
			if (rule) {
				results.splice(
					results.length,
					0,
					...rule.validateNode(node, variables, namespaceResolver)
				);
			}

			return Array.from(node.childNodes).reduce(flattenValidationResults, results);
		};

		return Array.from(documentDom.childNodes).reduce(flattenValidationResults, []);
	}

	static QUERY = `map {
		'id': @id/string(),
		'rules': array{ ./sch:rule/${Rule.QUERY}},
		'variables': array { ./sch:let/${Variable.QUERY}}
	}`;

	static fromJson(json: JsonPattern): Pattern {
		return new Pattern(
			json.id,
			json.rules.map(obj => Rule.fromJson(obj)),
			json.variables.map(obj => Variable.fromJson(obj))
		);
	}
}

export type JsonPattern = {
	id: string | null;
	rules: JsonRule[];
	variables: JsonVariable[];
};
