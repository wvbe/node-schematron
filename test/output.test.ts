import { Schema } from '../src/index';

const SCHEMATRON_EXAMPLE = `<schema xmlns="http://purl.oclc.org/dsdl/schematron">
	<ns prefix="drp" uri="http://derp" />
	<pattern>
		<rule context="drp:thunder[@foo]">
			<report test="true()">
				A report that always fires <name /> for any foo
			</report>
			<report test="false()" id="fires-never">
				A report that never fires <name /> for any foo
			</report>
			<assert test="not(@foo = 'bar' and not(preceding-sibling::*))" id="foobar-1st-child">
				An assert that fails once out of three times
			</assert>
		</rule>
		<rule context="drp:thunder">
			<assert test="@foo = 'bar'" id="fires-always">
				An assert that applies to only one element and always fails
			</assert>
		</rule>
	</pattern>
</schema>`;

const DOCUMENT_EXAMPLE = `<xml xmlns="http://derp" foo="err">
	<thunder />
	<thunder foo="unbar">
		<thunder foo="bar" />
	</thunder>
	<thunder foo="bar" />
</xml>`;

describe('Output', () => {
	const schema = Schema.fromString(SCHEMATRON_EXAMPLE.replace(/\t|\n/g, ''));
	const results = schema.validateString(DOCUMENT_EXAMPLE.replace(/\t|\n/g, ''));

	it('Schema.fromString(SCHEMATRON_EXAMPLE)', () => {
		expect(schema).toMatchSnapshot();
	});

	it('schema.validateString(DOCUMENT_EXAMPLE)', () => {
		expect(results).toMatchSnapshot();
	});
});
