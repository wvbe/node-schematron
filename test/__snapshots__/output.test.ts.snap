// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output Schema.fromString(SCHEMATRON_EXAMPLE) 1`] = `
Schema {
  "defaultPhase": null,
  "namespaces": Array [
    Namespace {
      "prefix": "drp",
      "uri": "http://derp",
    },
  ],
  "patterns": Array [
    Pattern {
      "id": null,
      "rules": Array [
        Rule {
          "asserts": Array [
            Assert {
              "id": null,
              "isReport": true,
              "message": Array [
                "A report that always fires ",
                Object {
                  "$type": "name",
                },
                " for any foo",
              ],
              "test": "true()",
            },
            Assert {
              "id": "fires-never",
              "isReport": true,
              "message": Array [
                "A report that never fires ",
                Object {
                  "$type": "name",
                },
                " for any foo",
              ],
              "test": "false()",
            },
            Assert {
              "id": "foobar-1st-child",
              "isReport": false,
              "message": Array [
                "An assert that fails once out of three times",
              ],
              "test": "not(@foo = 'bar' and not(preceding-sibling::*))",
            },
          ],
          "context": "drp:thunder[@foo]",
          "variables": Array [],
        },
        Rule {
          "asserts": Array [
            Assert {
              "id": "fires-always",
              "isReport": false,
              "message": Array [
                "An assert that applies to only one element and always fails",
              ],
              "test": "@foo = 'bar'",
            },
          ],
          "context": "drp:thunder",
          "variables": Array [],
        },
      ],
      "variables": Array [],
    },
  ],
  "phases": Array [],
  "title": null,
  "variables": Array [],
}
`;

exports[`Output schema.validateString(DOCUMENT_EXAMPLE) 1`] = `
Array [
  Result {
    "assertId": "fires-always",
    "context": <thunder />,
    "isReport": false,
    "message": "An assert that applies to only one element and always fails",
  },
  Result {
    "assertId": null,
    "context": <thunder
      foo="unbar"
    >
      <thunder
        foo="bar"
      />
    </thunder>,
    "isReport": true,
    "message": "A report that always fires thunder for any foo",
  },
  Result {
    "assertId": null,
    "context": <thunder
      foo="bar"
    />,
    "isReport": true,
    "message": "A report that always fires thunder for any foo",
  },
  Result {
    "assertId": "foobar-1st-child",
    "context": <thunder
      foo="bar"
    />,
    "isReport": false,
    "message": "An assert that fails once out of three times",
  },
  Result {
    "assertId": null,
    "context": <thunder
      foo="bar"
    />,
    "isReport": true,
    "message": "A report that always fires thunder for any foo",
  },
]
`;
